### PR TITLE
Add validator-info CLI

### PR DESCRIPTION
**Pull request by CriesofCarrots**
Validators (for TdS and beyond) need a way to publish information about themselves.
This PR offers as a solution a simple registration CLI, using the Config on-chain program to store ValidatorInfo.

Validators are responsible for generating a `validator-info` keypair and passing it to the CLI as the registration info account location and signer.
The Validator Pubkey is also used to sign the ValidatorInfo, and a non-signer key (`Va1idator1nfo111111111111111111111111111111`) is included in the account data to facilitate client-side account filtering.

The CLI offers the following ValidatorInfo fields:

- Validator name
- Validator website
- Validator keybase id
- Validator details

Replaces #4942

### CODE REVIEW

**Comment by garious** [comment]
Very cool, but this front line app seems like an especially bad place to have no tests.

File: register/src/register.rs [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+
+    let (message, signers): (Message, [&Keypair; 2]) = match matches.subcommand() {
+        ("new", Some(matches)) => {
+            let validator_info = ValidatorInfo {
+                info: parse_args(&matches)?,
+            };
+            println!("Registering Validator {:?}", validator_keypair.pubkey());
+            let instructions = vec![
+                config_instruction::create_account::<ValidatorInfo>(
+                    &validator_keypair.pubkey(),
```
Comment (rob-solana) [inline register/src/register.rs]: is this the node keypair?

Comment (rob-solana) [reply register/src/register.rs]: nevermind, stupid Q

File: register/src/register.rs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+                    Arg::with_name("website")
+                        .short("w")
+                        .long("website")
+                        .value_name("URL")
+                        .takes_value(true)
+                        .validator(check_url)
+                        .help("Validator website url"),
+                )
+                .arg(
+                    Arg::with_name("keybase")
```
Comment (mvines) [inline register/src/register.rs]: nit: `keybase_id`?

**Comment by rob-solana** [comment]
I had to go look at config_api again to understand what's going on here and ended up with a list of concerns for config_api ;)

**Comment by mvines** [comment]
> I had to go look at config_api again to understand what's going on here and ended up with a list of concerns for config_api ;)

Oh like what?

File: register/src/register.rs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+solana_sdk::solana_name_id!(
+    REGISTER_CONFIG_KEY,
+    "Va1idator1nfo111111111111111111111111111111"
+);
+
+#[derive(Debug, Serialize)]
+struct ValidatorInfo {
+    #[serde(with = "key_value_vec")]
+    info: Vec<String>,
```
Comment (rob-solana) [inline register/src/register.rs]: does this form make it easy for clients?  would a simple string suffice?

File: register/src/register.rs [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+                seq.serialize_element(&byte)?;
+            }
+        }
+        seq.end()
+    }
+}
+
+impl ConfigState for ValidatorInfo {
+    fn max_space() -> u64 {
+        512
```
Comment (rob-solana) [inline register/src/register.rs]: 6 lines of text...

File: register/src/register.rs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        )
+        .get_matches();
+
+    let json_rpc_url = matches.value_of("json_rpc_url").unwrap();
+
+    let mut path = dirs::home_dir().expect("home directory");
+    let id_path = if matches.is_present("validator_keypair") {
+        matches.value_of("validator_keypair").unwrap()
+    } else {
+        path.extend(&[".config", "solana", "validator-keypair.json"]);
```
Comment (mvines) [inline register/src/register.rs]: Who creates `~/.config/solana/validator-keypair.json`?

File: register/src/register.rs [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        }
+        path.to_str().unwrap()
+    };
+    let validator_keypair = read_keypair(id_path)?;
+
+    let mut path = dirs::home_dir().expect("home directory");
+    let id_path = if matches.is_present("info_keypair") {
+        matches.value_of("info_keypair").unwrap()
+    } else {
+        path.extend(&[".config", "solana", "validator-info.json"]);
```
Comment (mvines) [inline register/src/register.rs]: Who creates `~/.config/solana/validator-info.json`?

**Comment by rob-solana** [comment]
> > I had to go look at config_api again to understand what's going on here and ended up with a list of concerns for config_api ;)
> 
> Oh like what?

I can put 'em here, but I should probably file issues...

* deserialize().unwrap() (crashes node)
* keyed_accounts[index] from a deserialized message (crashes node)
* the need to present config keys in the message in the same order as in the account data, which looks brittle
* if the account creator can say who has to sign to change the config, should we require that the account pubkey be forever among that set?

File: register/src/register.rs [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+            exit(1);
+        }
+        path.to_str().unwrap()
+    };
+    let info_keypair = read_keypair(id_path)?;
+
+    let keys = vec![(id(), false), (validator_keypair.pubkey(), true)];
+
+    let (message, signers): (Message, [&Keypair; 2]) = match matches.subcommand() {
+        ("new", Some(matches)) => {
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply register/src/register.rs]: I was wondering about this. The down side to that approach is that it involves an extra RPC call to check for the existence of the info account. Acceptable cost?

File: register/src/register.rs [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+            println!("No info keypair file found. Run solana-keygen to create one.");
+            exit(1);
+        }
+        path.to_str().unwrap()
+    };
+    let info_keypair = read_keypair(id_path)?;
+
+    let keys = vec![(id(), false), (validator_keypair.pubkey(), true)];
+
+    let (message, signers): (Message, [&Keypair; 2]) = match matches.subcommand() {
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply register/src/register.rs]: Yep, on my future radar as well. Might squeeze in now, if I'm reworking the args anyway.

File: register/src/register.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mvines) [reply register/src/register.rs]: > Acceptable cost?

IMO yep

File: register/src/register.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (mvines) [reply register/src/register.rs]: `get --all` too?  ;)

File: register/src/register.rs (hunk 4 above)
Comment (CriesofCarrots) [reply register/src/register.rs]: @rob-solana , are you suggesting that's too much or too little? Or...?

File: register/src/register.rs (hunk 5 above)
Comment (CriesofCarrots) [reply register/src/register.rs]: This is the validator node keypair, which we walk validators through creating in the testnet participation guide.

File: register/src/register.rs (hunk 6 above)
Comment (CriesofCarrots) [reply register/src/register.rs]: Currently, "Validators are responsible for generating a validator-info keypair and passing it to the CLI as the registration info account location and signer." (from PR description)
We could have the validator-info script generate that keypair, but that seems awkward from a real-world key-management perspective.

File: register/src/register.rs [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+                        .help("Validator Keybase id"),
+                )
+                .arg(
+                    Arg::with_name("details")
+                        .short("d")
+                        .long("details")
+                        .value_name("STRING")
+                        .takes_value(true)
+                        .validator(check_details_length)
+                        .help("Validator description, max characters: ###"),
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply register/src/register.rs]: It should... but needs updating.

File: register/src/register.rs (hunk 6 above)
Comment (mvines) [reply register/src/register.rs]: Got it.  Can you add the extra steps to the testnet participation guide

Comment (CriesofCarrots) [reply register/src/register.rs]: Will do!

File: register/src/register.rs (hunk 3 above)
Comment (CriesofCarrots) [reply register/src/register.rs]: Not easier, but I think not harder. And it takes up less space.

Comment (CriesofCarrots) [reply register/src/register.rs]: Will go that direction instead.

Comment (mvines) [reply register/src/register.rs]: msgpack is nice: https://crates.io/crates/msgpack

File: register/src/register.rs (hunk 4 above)
Comment (rob-solana) [reply register/src/register.rs]: seems small to me

File: validator-info/src/validator_info.rs [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+                if !path.exists() {
+                    println!("No validator keypair file found. Run solana-keygen to create one.");
+                    exit(1);
+                }
+                path.to_str().unwrap()
+            };
+            let validator_keypair = read_keypair(id_path)?;
+
+            // Load validator-info keypair
+            let mut path = dirs::home_dir().expect("home directory");
```
Comment (mvines) [inline validator-info/src/validator_info.rs]: Couldn't we make the validator info keypair ephemeral?  It never needs to hit disk, it only needs to exist long enough to transfer update access to the validator's keypair.

Comment (CriesofCarrots) [reply validator-info/src/validator_info.rs]: Yep, reworked

File: validator-info/src/validator_info.rs [hunk 11]
```diff
@@ -0,0 +1,506 @@
...
+            MAX_SHORT_FIELD_LENGTH
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+// Return an error if invalid keybase id is submitted.
+fn is_keybase_id(string: String) -> Result<(), String> {
+    let mut invalid = string.len() != 16;
```
Comment (mvines) [inline validator-info/src/validator_info.rs]: A keybase user id does not need to be exactly 16 characters, mine is 5 ([mvines](https://keybase.io/mvines)].  If one adds a PGP key to their keybase account, that is 16 characters, but a GPG key is not mandatory for an account and it looks like it's hard to search for a user based on a PGP key with the keybase APIs (whereas searching by user id is trivial).

File: validator-info/src/validator_info.rs [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+                path.extend(&[".config", "solana", "validator-keypair.json"]);
+                if !path.exists() {
+                    println!("No validator keypair file found. Run solana-keygen to create one.");
+                    exit(1);
+                }
+                path.to_str().unwrap()
+            };
+            let validator_keypair = read_keypair(id_path)?;
+
+            // Create validator-info keypair to use if info_pubkey no provided or DNE
```
Comment (mvines) [inline validator-info/src/validator_info.rs]: "DNE"?

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Pull request merged**